newrelic/hooks/mlmodel_gemini.py

```diff
@@ -0,0 +1,560 @@
+from newrelic.api.transaction import current_transaction
+from newrelic.common.object_wrapper import wrap_function_wrapper
+from newrelic.common.package_version_utils import get_package_version
+from newrelic.core.config import global_settings
+
+GEMINI_VERSION = get_package_version("google-genai")
+EXCEPTION_HANDLING_FAILURE_LOG_MESSAGE = (
+    "Exception occurred in Gemini instrumentation: While reporting an exception "
+    "in Gemini, another exception occurred. Report this issue to New Relic "
+    "Support.\n "
+)
+RECORD_EVENTS_FAILURE_LOG_MESSAGE = (
+    "Exception occurred in Gemini instrumentation: Failed to record LLM events. "
+    "Please report this issue to New Relic Support.\n "
+)
+
+_logger = logging.getLogger(__name__)
+
+
+def wrap_embed_content_sync(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return wrapped(*args, **kwargs)
+
+    settings = transaction.settings or global_settings()
+    if not settings.ai_monitoring.enabled:
+        return wrapped(*args, **kwargs)
+
+    # Framework metric also used for entity tagging in the UI
+    transaction.add_ml_model_info("Gemini", GEMINI_VERSION)
+    transaction._add_agent_attribute("llm", True)
+
+    # Obtain attributes to be stored on embedding events regardless of whether we hit an error
+    embedding_id = str(uuid.uuid4())
+
+    ft = FunctionTrace(name=wrapped.__name__, group="Llm/embedding/Gemini")
+    ft.__enter__()
+    linking_metadata = get_trace_linking_metadata()
+    try:
+        response = wrapped(*args, **kwargs)
+    except Exception as exc:
+        _record_embedding_error(transaction, embedding_id, linking_metadata, kwargs, ft, exc)
+        raise
+    ft.__exit__(None, None, None)
+
+    if not response:
+        return response
+
+    _record_embedding_success(transaction, embedding_id, linking_metadata, kwargs, ft)
+    return response
+
+
+async def wrap_embed_content_async(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return await wrapped(*args, **kwargs)
+
+    settings = transaction.settings or global_settings()
+    if not settings.ai_monitoring.enabled:
+        return await wrapped(*args, **kwargs)
+
+    # Framework metric also used for entity tagging in the UI
+    transaction.add_ml_model_info("Gemini", GEMINI_VERSION)
+    transaction._add_agent_attribute("llm", True)
+
+    # Obtain attributes to be stored on embedding events regardless of whether we hit an error
+    embedding_id = str(uuid.uuid4())
+
+    ft = FunctionTrace(name=wrapped.__name__, group="Llm/embedding/Gemini")
+    ft.__enter__()
+    linking_metadata = get_trace_linking_metadata()
+    try:
+        response = await wrapped(*args, **kwargs)
+    except Exception as exc:
+        _record_embedding_error(transaction, embedding_id, linking_metadata, kwargs, ft, exc)
+        raise
+    ft.__exit__(None, None, None)
```

Comment:
Same thing here, could use a comment for this. And wherever it came from could as well.